### PR TITLE
Fix : #58748 BUG: eval fails for ExtensionArray

### LINKED ISSUE #58748: BUG: eval fails for ExtensionArray
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd

df = pd.DataFrame({'a': pd.array([1, 2, 3]), 'b': pd.array([4, 5, 6])})

# this works as expected
print(df['a'] / df['b'])
"""
0    0.25
1     0.4
2     0.5
dtype: Float64
"""

# this is not working
print(df.eval('a / b'))
"""
TypeError: Cannot interpret 'Int64Dtype()' as a data type
"""
```

### Issue Description

This issue is coming from [pint-pandas#137](https://github.com/hgrecco/pint-pandas/issues/137). `ExtensionArray` is used for this project and `df.eval` inlcuding division (`\`) fails with `TypeError`.

I was able to able to reproduce the pint-pandas  issue with pandas build in `IntegerArray` as shown in the above example 

<details>

``` python-traceback
Traceback (most recent call last):
  File "c:\CALC\DEV\pint pandas debug\test.py", line 17, in <module>
    print(df.eval('a / b'))
          ^^^^^^^^^^^^^^^^
  File "C:\Users\laurent.mutricy\AppData\Local\miniconda3\Lib\site-packages\pandas\core\frame.py", line 4738, in eval
    return _eval(expr, inplace=inplace, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\laurent.mutricy\AppData\Local\miniconda3\Lib\site-packages\pandas\core\computation\eval.py", line 340, in eval
    parsed_expr = Expr(expr, engine=engine, parser=parser, env=env)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\laurent.mutricy\AppData\Local\miniconda3\Lib\site-packages\pandas\core\computation\expr.py", line 809, in __init__
    self.terms = self.parse()
                 ^^^^^^^^^^^^
  File "C:\Users\laurent.mutricy\AppData\Local\miniconda3\Lib\site-packages\pandas\core\computation\expr.py", line 828, in parse
    return self._visitor.visit(self.expr)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\laurent.mutricy\AppData\Local\miniconda3\Lib\site-packages\pandas\core\computation\expr.py", line 413, in visit
    return visitor(node, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\laurent.mutricy\AppData\Local\miniconda3\Lib\site-packages\pandas\core\computation\expr.py", line 419, in visit_Module
    return self.visit(expr, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\laurent.mutricy\AppData\Local\miniconda3\Lib\site-packages\pandas\core\computation\expr.py", line 413, in visit
    return visitor(node, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\laurent.mutricy\AppData\Local\miniconda3\Lib\site-packages\pandas\core\computation\expr.py", line 422, in visit_Expr
    return self.visit(node.value, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\laurent.mutricy\AppData\Local\miniconda3\Lib\site-packages\pandas\core\computation\expr.py", line 413, in visit
    return visitor(node, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\laurent.mutricy\AppData\Local\miniconda3\Lib\site-packages\pandas\core\computation\expr.py", line 535, in visit_BinOp
    return self._maybe_evaluate_binop(op, op_class, left, right)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\laurent.mutricy\AppData\Local\miniconda3\Lib\site-packages\pandas\core\computation\expr.py", line 502, in _maybe_evaluate_binop
    res = op(lhs, rhs)
          ^^^^^^^^^^^^
  File "C:\Users\laurent.mutricy\AppData\Local\miniconda3\Lib\site-packages\pandas\core\computation\expr.py", line 538, in <lambda>
    return lambda lhs, rhs: Div(lhs, rhs)
                            ^^^^^^^^^^^^^
  File "C:\Users\laurent.mutricy\AppData\Local\miniconda3\Lib\site-packages\pandas\core\computation\ops.py", line 528, in __init__
    if not isnumeric(lhs.return_type) or not isnumeric(rhs.return_type):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\laurent.mutricy\AppData\Local\miniconda3\Lib\site-packages\pandas\core\computation\ops.py", line 512, in isnumeric
    return issubclass(np.dtype(dtype).type, np.number)
                      ^^^^^^^^^^^^^^^
TypeError: Cannot interpret 'Int64Dtype()' as a data type
```

</details>

Looking at the traceback it seams that  `pandas\core\computation\ops.isnumeric()` does not handle properly dtypes from `ExtensionArray` class (or subclass). It was suggested to consider `_is_numeric` class attribute. 

```diff
def isnumeric(dtype) -> bool:
-    return issubclass(np.dtype(dtype).type, np.number)
+   return getattr(dtype, '_is_numeric', False) or issubclass(np.dtype(dtype).type, np.number)
```

probably linked to #21374

### Expected Behavior

`df.eval('a / b')` should give the same result than `df['a'] / df['b']`

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit                : 4fb94bb09abbfcae536744740b00c987a04eed23
python                : 3.11.4.final.0
python-bits           : 64
OS                    : Windows
OS-release            : 10
Version               : 10.0.19045
machine               : AMD64
processor             : Intel64 Family 6 Model 186 Stepping 2, GenuineIntel
byteorder             : little
LC_ALL                : None
LANG                  : en_US.UTF-8
LOCALE                : fr_FR.cp1252

pandas                : 3.0.0.dev0+983.g4fb94bb09a
numpy                 : 1.26.4
pytz                  : 2024.1
dateutil              : 2.8.2
setuptools            : 67.8.0
pip                   : 23.1.2
Cython                : None
pytest                : 8.2.0
hypothesis            : None
sphinx                : None
blosc                 : None
feather               : None
xlsxwriter            : None
lxml.etree            : None
html5lib              : None
pymysql               : None
psycopg2              : None
jinja2                : None
IPython               : None
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : None
bottleneck            : None
fastparquet           : None
fsspec                : None
gcsfs                 : None
matplotlib            : 3.8.3
numba                 : None
numexpr               : 2.10.0
odfpy                 : None
openpyxl              : 3.1.2
pyarrow               : 15.0.0
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : 1.12.0
sqlalchemy            : None
tables                : None
tabulate              : 0.9.0
xarray                : 2024.2.0
xlrd                  : None
zstandard             : 0.19.0
tzdata                : 2023.3
qtpy                  : None
pyqt5                 : None

</details>

### PR DESCRIPTION
**Pull request by SiddheshBangar**
- [x] closes #58748  (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/computation/ops.py [hunk 1]
```diff
@@ -509,7 +509,7 @@ def _disallow_scalar_only_bool_ops(self) -> None:
 
 
 def isnumeric(dtype) -> bool:
-    return issubclass(np.dtype(dtype).type, np.number)
+    return getattr(dtype, '_is_numeric', False) or issubclass(np.dtype(dtype).type, np.number)
```
Comment (mutricyl) [inline pandas/core/computation/ops.py:512]:
```suggestion
        return getattr(dtype, "_is_numeric", False) or issubclass(
        np.dtype(dtype).type, np.number
    )
```
This is needed for `pre-commit` to success

**Review by mutricyl — COMMENTED** [review]
Would you consider adding a test ? maybe based on the example I provided ?
it may be located in `pandas\tests\extension\test_datetime.py`

**Comment by SiddheshBangar** [comment]
Hey @mutricyl thanks for mentioning it, I have worked on it and did the necessary changes. Let me know if any other changes need to be made I will work on it. Thanks

File: pandas/tests/extension/test_datetime.py [hunk 2]
```diff
@@ -88,6 +88,13 @@ def cmp(a, b):
     return cmp
 
 
+@pytest.fixture
+def isnumeric(dtype):
+    return getattr(dtype, "is_numeric", False) or issubclass(
+        np.dtype(dtype).type, np.number
+    )
```
Comment (rhshadrach) [inline pandas/tests/extension/test_datetime.py:95]: pytest fixtures aren't themselves tests, they are used to help in the creation of tests. In other words, this alone does not test anything. The test should likely go in `tests/frame/test_query_eval.py`.

File: pandas/core/computation/ops.py [hunk 3]
```diff
@@ -509,7 +509,9 @@ def _disallow_scalar_only_bool_ops(self) -> None:
 
 
 def isnumeric(dtype) -> bool:
```
Comment (rhshadrach) [inline pandas/core/computation/ops.py:511]: Can we replace this with `pandas.core.dtypes.common.is_numeric_dtype`.

Comment (mutricyl) [reply pandas/core/computation/ops.py:511]: `pandas.core.dtypes.common.is_numeric_dtype` looks suitable. `isnumeric` function may be removed and `__init__` of `class Div(BinOp)` may be changed as follow:

```diff
-        if not isnumeric(lhs.return_type) or not isnumeric(rhs.return_type):
+        if not is_numeric_dtype(lhs.return_type) or not is_numeric_dtype(
+            rhs.return_type
+        ):
```
( `is_numeric_dytpe` added in imports of course)

@SiddheshBangar test may look like this:

``` python
def test_extension_array_eval():
    # GH#58748
    df = DataFrame({"a": pd.array([1, 2, 3]), "b": pd.array([4, 5, 6])})
    result = df.eval("a / b")
    expected = Series([0.25, 0.40, 0.50])
    tm.assert_series_equal(result, expected)
```

**Comment by SiddheshBangar** [comment]
Thanks @rhshadrach @mutricyl I will work on this and update the PR as well

**Comment by SiddheshBangar** [comment]
Although, I tried making necessary changes, and I run the pre-commits as well didn't faced any issues in it but while doing here some checks are failing any guidance how can I move forward to this?

File: pandas/tests/frame/test_query_eval.py [hunk 4]
```diff
@@ -202,6 +202,13 @@ def test_eval_simple(self, engine, parser):
         expected = df["a"]
         tm.assert_series_equal(expected, res)
 
+    def test_extension_array_eval():
```
Comment (rhshadrach) [inline pandas/tests/frame/test_query_eval.py:205]: Where this is indented, it is part of a class. It needs `self` as a first argument.

Comment (mutricyl) [inline pandas/tests/frame/test_query_eval.py:205]: You have indented your test function as if it was a class method. That is likely why test are failing.

**Comment by SiddheshBangar** [comment]
@rhshadrach @mutricyl pushed one more PR, I fixed the issues, sorry from my end once again. Let me know if this works perfectly. Once again thanks for the help.

**Comment by SiddheshBangar** [comment]
Hey @rhshadrach @mutricyl just a follow up message that could you check the PR that I have pushed a while ago I hope this works for your guys do check and let me know if any more changes are required to be done from my end as well. Thanks!

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Looks good - can you add a line in the whatsnew for 3.0 in the `Other` section. That section mentions not to use it, but I believe this case to be appropriate.

File: pandas/tests/frame/test_query_eval.py [hunk 5]
```diff
@@ -203,6 +203,15 @@ def test_eval_simple(self, engine, parser):
         tm.assert_series_equal(expected, res)
 
 
+class TestDataFrameEvalWithMultiIndex:
```
Comment (andrewgsavage) [inline pandas/tests/frame/test_query_eval.py:206]: This test doesn't use a MultiIndex. I'd keep it part of `TestDataFrameEval`

Comment (SiddheshBangar) [reply pandas/tests/frame/test_query_eval.py:206]: Okay works

File: doc/source/whatsnew/v3.0.0.rst [hunk 6]
```diff
@@ -593,6 +593,7 @@ Styler
 Other
 ^^^^^
 - Bug in :class:`DataFrame` when passing a ``dict`` with a NA scalar and ``columns`` that would always return ``np.nan`` (:issue:`57205`)
+- Bug in :func:`eval` including division ``\`` failed with a ``TypeError``. (:issue:`58748`)
```
Comment (rhshadrach) [inline doc/source/whatsnew/v3.0.0.rst:596]: nit: division is `/`, not `\`. Can you also include it's with `` :class:`ExtensionArray` `` dtypes.

Comment (SiddheshBangar) [reply doc/source/whatsnew/v3.0.0.rst:596]: Okay got it, I will look into this, Thanks.

**Comment by SiddheshBangar** [comment]
@rhshadrach updated the docs and committed the changes do check the PR and let me know if this works, Thanks!

**Review by rhshadrach — APPROVED** [review]
lgtm

**Comment by rhshadrach** [comment]
Thanks @SiddheshBangar

**Comment by SiddheshBangar** [comment]
Welcome, thanks for helping and the guidance @rhshadrach, @mutricyl

**Pull request merged**